src/shared/locked-status/use-check-lock-status.js

```diff
@@ -10,96 +12,161 @@ import { useDataValueSet } from '../use-data-value-set/use-data-value-set.js'
+            selectors.getApplicableDataInputPeriod(
+                metadata,
+                dataSetId,
+                selectedPeriod.id
+            )
+
+        if (!applicableDataInputPeriod) {
+            // If there are defined data input periods, but there is not one
+            // for this selected period, then the form is closed
+            return {
+                state: LockedStates.LOCKED_DATA_INPUT_PERIOD,
+                lockDate: null,
+            }
+        }
+
+        const { openingDate, closingDate } = applicableDataInputPeriod
+        // openingDate and closingDate can be undefined.
+        // They are local to the server
+        const parsedOpeningDate =
+            openingDate && fromServerDate(new Date(openingDate))
+        const parsedClosingDate =
+            closingDate && fromServerDate(new Date(closingDate))
+
+        if (
+            (openingDate &&
+                currentDate.serverDate < parsedOpeningDate.serverDate) ||
+            (closingDate &&
+                currentDate.serverDate > parsedClosingDate.serverDate)
+        ) {
+            return {
+                state: LockedStates.LOCKED_DATA_INPUT_PERIOD,
+                lockDate: null,
+            }
+        }
 
-    if (!dataInputPeriod) {
-        return false
+        // If we're here, the form isn't (yet) locked by the data input period.
+        // Set the clientLockDate to the input period's closing date (if this
+        // input period has one) before also checking expiry days.
+        // This might still be undefined, but that's okay
+        clientLockDate = parsedClosingDate?.clientDate
     }
 
-    const currentDateAtServerTimeZone = new Date(currentDayString)
-    const openingDate = new Date(dataInputPeriod.openingDate)
-    const closingDate = new Date(dataInputPeriod.closingDate)
+    if (expiryDays > 0) {
+        // selectedPeriod.endDate is a string like '2023-02-20' --
+        // convert that to a date and get time to get ms to add expiry days.
+        // Also add one day more because selectedPeriod.endDate is the START
+        // of the period's last day (00:00), and we want the end of that day
+        // (confirmed with backend behavior).
+        // Converting the selectedPeriod.endDate string to an object creates
+        // it in the client's timezone; parse it from there.
+        const expiryDate = fromClientDate(
+            new Date(
+                new Date(selectedPeriod.endDate).getTime() +
+                    (expiryDays + 1) * DAY_MS
+            )
+        )
 
-    return (
-        openingDate > currentDateAtServerTimeZone ||
-        closingDate < currentDateAtServerTimeZone
-    )
+        if (currentDate.serverDate < expiryDate.serverDate) {
+            // Take the sooner of the two possible lock dates
+            clientLockDate = clientLockDate
+                ? new Date(Math.min(clientLockDate, expiryDate.clientDate))
+                : expiryDate.clientDate
```

Comment:
I guess you could have logic here about the user so that if they are superuser (have ALL authorities), you would not record a clientLockDate that is associated with expiryDays? (e.g. so you don't display the info about it closing if it's because of expiryDays and they're a superuser) (we're already getting the authorities, so I guess adding a check for ALL authorities would be pretty minor).

Reply:
Yeah that's pretty easy -- added it 🙂